### PR TITLE
fix #3 - implement service system

### LINKED ISSUE #3: create message queue module
We need to propose the syntax.

### PR DESCRIPTION
**Pull request by john-hu**

### CODE REVIEW

**Comment by john-hu** [comment]
wait for testing

File: js/services.js [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+'use strict';
+
+(function(exports) {
+  /**
+   * This module is responsible to create and run services.
+   */
+  var Services = function() {};
+  Services.prototype = {
+
+    _dispatchReady: function() {
```
Comment (gasolin) [inline js/services.js:12]: it's really hard to name it, How about with full event name `_dispatchSystemMessageListenerReady`

File: js/services.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+      this._serviceMap = {};
+      this._serviceContainer = document.getElementById('service-section');
+      this._windowContainer = document.getElementById('window-section');
+      // We need to be sure to get the focus in order to wake up the screen
+      // if the phone goes to sleep before any user interaction.
+      // Apparently it works because no other window
+      // has the focus at this point.
+      window.focus();
+
+      this._clonedConfig = [].concat(ServiceConfig);
```
Comment (gasolin) [inline js/services.js:160]: The ServiceConfig is actually the rc.d like list right? why not call it `initialLaunchList` or something

File: js/services.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+      // mozbrowserloadstart, which sometimes does not happen till
+      // window.onload.
+      var evt = new CustomEvent('mozContentEvent', {
+        bubbles: true, cancelable: false,
+        detail: { type: 'system-message-listener-ready'}
+      });
+      window.dispatchEvent(evt);
+    },
+
+    handleBrowserLoaded: function(instance) {
```
Comment (gasolin) [inline js/services.js:26]: It's better go with basic doc that what is window/app for -> or daemon/module/app for , ex:

```
there are 3 types of container provided by system: daemon, module, app
* daemon: module runs without visual presentation
* module: embedded iframe, can communicate with system with postMessage
* app: module runs in different process
```

File: js/services.js [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+      }
+
+      if (instance.config.type === 'app') {
+        // We cannot create messenger for an app since it is at remote
+        // process.
+        // move focus to foreground window
+        document.activeElement.blur();
+        instance.iframe.focus();
+        // process next one
+        this._processCreateQueue();
```
Comment (gasolin) [inline js/services.js:40]: how about `_initNextContainer()`

File: js/services.js [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      }.bind(this);
+    },
+
+    handleBrowserError: function(instance) {
+      if (instance.config.type === 'window' ||
+          instance.config.type === 'app') {
+        // move focus to top most window
+        document.activeElement.blur();
+        var iframe = this._windowContainer.querySelector(
+                                                       'iframe:last-child');
```
Comment (gasolin) [inline js/services.js:66]: why change line like this @@

File: js/services.js [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+      instance.messenger.stop();
+
+      instance.iframe.parentElement.removeChild(instance.iframe);
+      if (instance.config.autoRestart) {
+        this._createService(instance.config);
+      }
+    },
+
+    handleEvent: function(e) {
+      var instance = this._serviceMap[e.target.getAttribute('mozapp')];
```
Comment (gasolin) [inline js/services.js:81]: wish oneday we need to wrap service mapping to another method :dancer: 
Its totally fine now.

File: js/services.js [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    },
+
+    _processCreateQueue: function() {
+      if (this._clonedConfig.length === 0) {
+        this.send({'log': 'all services created'});
+        delete this._clonedConfig;
+        this._dispatchReady();
+        return;
+      }
+      this._createService(this._clonedConfig.pop());
```
Comment (gasolin) [inline js/services.js:99]: how about `_createContainer`

File: js/services.js [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+    },
+
+    _createService: function(s) {
```
Comment (gasolin) [inline js/services.js:103]: rename s to `manifest` or `config` might be more clear

File: js/services.js (hunk 3 above)
Comment (john-hu) [inline js/services.js:26]: Sure. We should write it at our wiki.

File: js/services.js (hunk 4 above)
Comment (john-hu) [inline js/services.js:40]: container is too abstract. I will change to Daemon.

File: js/services.js (hunk 5 above)
Comment (john-hu) [inline js/services.js:66]: 80 lines.

File: js/services.js (hunk 7 above)
Comment (john-hu) [inline js/services.js:99]: keep it for renaming work.

File: js/services.js (hunk 8 above)
Comment (john-hu) [inline js/services.js:103]: Sure. I would call it as service at first. But it may be conflict with es6.

File: js/services.js (hunk 2 above)
Comment (john-hu) [inline js/services.js:160]: I don't have a good name now. And initialLaunchList is improper, too. I will keep it for renaming issue.

**Pull request merged**